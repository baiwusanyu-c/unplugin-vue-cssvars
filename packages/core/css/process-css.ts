import path from 'path'
import * as csstree from 'css-tree'
import { completeSuffix, transformSymbol } from '@unplugin-vue-cssvars/utils'
import { walkCSSTree } from './pre-process-css'
import type { ICSSFile, ICSSFileMap } from '../types'
import type { SFCDescriptor } from '@vue/compiler-sfc'

export const getCSSFileRecursion = (key: string, cssFiles: ICSSFileMap, cb: (res: ICSSFile) => void, matchedMark = new Set<string>()) => {
  // 避免循环引用
  if (matchedMark.has(key)) return
  const cssFile = cssFiles.get(key)
  if (cssFile) {
    matchedMark.add(key)
    cb(cssFile)
    if (cssFile.importer.size > 0) {
      cssFile.importer.forEach((value) => {
        getCSSFileRecursion(value, cssFiles, cb, matchedMark)
      })
    }
  }
}

/**
 * 遍历 sfc 的 style 标签内容
 * 根据其 ast，获取 @import 信息
 * @param descriptor
 * @param id transform's id
 * @param cssFiles
 */
export const createCSSModule = (descriptor: SFCDescriptor, id: string, cssFiles: ICSSFileMap) => {
  const importModule: Array<ICSSFile> = []
  // 遍历 sfc 的 style 标签内容
  for (let i = 0; i < descriptor.styles.length; i++) {
    const content = descriptor.styles[i].content
    const cssAst = csstree.parse(content)
    // 根据其 ast，获取 @import 信息
    walkCSSTree(cssAst, (importer) => {
      // 添加后缀
      // sfc中规则：如果@import 指定了后缀，则根据后缀，否则根据当前 script 标签的 lang 属性（默认css）
      let key = completeSuffix(transformSymbol(path.resolve(path.parse(id).dir, importer)), descriptor.styles[i].lang)
      // 如果 .scss 的 import 不存在，则用 css 的
      if (!cssFiles.get(key))
        key = completeSuffix(transformSymbol(path.resolve(path.parse(id).dir, importer)))

      // 根据 @import 信息，从 cssFiles 中，递归的获取所有在预处理时生成的 cssvars 样式
      getCSSFileRecursion(key, cssFiles, (res: ICSSFile) => {
        importModule.push(res)
      })
    }, { i: true, v: false })
  }
  return importModule
}
