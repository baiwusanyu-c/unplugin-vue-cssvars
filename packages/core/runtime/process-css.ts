import { parse, resolve } from 'path'
import { SUPPORT_FILE, completeSuffix, setTArray } from '@unplugin-vue-cssvars/utils'
import { parseImports } from '../parser'
import type { ICSSFile, ICSSFileMap } from '../types'
import type { SFCDescriptor } from '@vue/compiler-sfc'

export const getCSSFileRecursion = (
  key: string,
  cssFiles: ICSSFileMap,
  cb: (res: ICSSFile) => void,
  matchedMark = new Set<string>()) => {
  // 避免循环引用
  if (matchedMark.has(key)) return
  const cssFile = cssFiles.get(key)
  if (cssFile) {
    matchedMark.add(key)
    cb(cssFile)
    if (cssFile.importer.size > 0) {
      cssFile.importer.forEach((value) => {
        getCSSFileRecursion(value, cssFiles, cb, matchedMark)
      })
    }
  }
}

/**
 * 遍历 sfc 的 style 标签内容
 * 根据其 ast，获取 @import 信息
 * @param descriptor
 * @param id transform's id
 * @param cssFiles
 * @param server
 */
// TODO: unit test
export type TInjectCSSContent = Set<{ content: string, lang: string, styleTagIndex: number }>
export const getVBindVariableListByPath = (
  descriptor: SFCDescriptor,
  id: string,
  cssFiles: ICSSFileMap,
  server: boolean) => {
  const vbindVariable: Set<string> = new Set()
  const injectCSSContent: TInjectCSSContent = new Set()
  // 遍历 sfc 的 style 标签内容
  for (let i = 0; i < descriptor.styles.length; i++) {
    const content = descriptor.styles[i].content
    const lang = descriptor.styles[i].lang === SUPPORT_FILE.STYLUS ? SUPPORT_FILE.STYL : descriptor.styles[i].lang
    const idDirParse = parse(id)
    const parseImporterRes = parseImports(content)
    parseImporterRes.imports.forEach((res) => {
      const importerPath = resolve(idDirParse.dir, res.path)
      // 添加后缀
      // sfc中规则：如果@import 指定了后缀，则根据后缀，否则根据当前 script 标签的 lang 属性（默认css）
      let key = completeSuffix(importerPath, lang)
      // 如果 .scss 的 import 不存在，则用 css 的
      if (!cssFiles.get(key))
        key = completeSuffix(importerPath)

      // 根据 @import 信息，从 cssFiles 中，递归的获取所有在预处理时生成的 cssvars 样式
      getCSSFileRecursion(key, cssFiles, (res: ICSSFile) => {
        if (res.vBindCode) {
          !server && injectCSSContent.add({ content: res.content, lang: res.lang, styleTagIndex: i })
          res.vBindCode.forEach((vb) => {
            vbindVariable.add(vb)
          })
        }
      })
    })
  }
  return {
    vbindVariableListByPath: setTArray(vbindVariable),
    injectCSSContent,
  }
}
