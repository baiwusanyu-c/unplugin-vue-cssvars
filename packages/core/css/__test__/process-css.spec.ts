import { describe, expect, test, vi } from 'vitest'
import { createCSSModule, getCSSFileRecursion } from '../process-css'
import type { ICSSFile } from '../../types'
describe('process css', () => {
  test('getCSSFileRecursion: basic', () => {
    const mockEvt = vi.fn()
    const mockRes: Array<ICSSFile> = []
    const mockCssFiles = new Map()
    mockCssFiles.set('foo', {
      importer: new Set(),
      vBindCode: {
        foo: new Set(['v-bind(foo)']),
      },
    })
    getCSSFileRecursion('foo', mockCssFiles, (v) => {
      mockEvt()
      mockRes.push(v)
    })
    expect(mockRes.length).toBe(1)
    expect(mockEvt).toBeCalledTimes(1)
    expect(mockRes[0].vBindCode).toMatchObject({
      foo: new Set(['v-bind(foo)']),
    })
  })

  test('getCSSFileRecursion: unmatched key', () => {
    const mockEvt = vi.fn()
    const mockRes = []
    const mockCssFiles = new Map()
    mockCssFiles.set('foo', {
      importer: new Set(['foo1', 'foo2']),
    })
    getCSSFileRecursion('bar', mockCssFiles, (v) => {
      mockEvt()
      mockRes.push(v)
    })
    expect(mockRes.length).toBe(0)
    expect(mockEvt).not.toBeCalled()
  })

  test('getCSSFileRecursion: recursion', () => {
    const mockEvt = vi.fn()
    const mockRes: Array<ICSSFile> = []
    const mockCssFiles = new Map()
    mockCssFiles.set('foo', {
      importer: new Set(['bar']),
      vBindCode: {
        foo: new Set(['v-bind(foo)']),
      },
    })
    mockCssFiles.set('bar', {
      importer: new Set(),
      vBindCode: {
        bar: new Set(['v-bind(bar)']),
      },
    })
    getCSSFileRecursion('foo', mockCssFiles, (v) => {
      mockEvt()
      mockRes.push(v)
    })
    expect(mockRes.length).toBe(2)
    expect(mockEvt).toBeCalledTimes(2)
    expect(mockRes[0].vBindCode).toMatchObject({
      foo: new Set(['v-bind(foo)']),
    })
    expect(mockRes[1].vBindCode).toMatchObject({
      bar: new Set(['v-bind(bar)']),
    })
  })

  test('getCSSFileRecursion: circular dependencies', () => {
    const mockEvt = vi.fn()
    const mockRes: Array<ICSSFile> = []
    const mockCssFiles = new Map()
    mockCssFiles.set('foo', {
      importer: new Set(['bar']),
      vBindCode: {
        foo: new Set(['v-bind(foo)']),
      },
    })
    mockCssFiles.set('bar', {
      importer: new Set(['foo']),
      vBindCode: {
        bar: new Set(['v-bind(bar)']),
      },
    })
    getCSSFileRecursion('foo', mockCssFiles, (v) => {
      mockEvt()
      mockRes.push(v)
    })
    expect(mockRes.length).toBe(2)
    expect(mockEvt).toBeCalledTimes(2)
    expect(mockRes[0].vBindCode).toMatchObject({
      foo: new Set(['v-bind(foo)']),
    })
    expect(mockRes[1].vBindCode).toMatchObject({
      bar: new Set(['v-bind(bar)']),
    })
  })

  test('createCSSModule: basic', () => {
    const mockCssFiles = new Map()
    const mockCSSFilesContent = {
      importer: new Set(),
      vBindCode: {
        foo: new Set(['v-bind(foo)']),
      },
    }
    mockCssFiles.set('D:/project-github/unplugin-vue-cssvars/play/src/assets/test.css', mockCSSFilesContent)
    const mockDescriptor = {
      styles: [{
        content: '@import "./assets/test";\n'
          + ' div {\n'
          + '   color: v-bind(color2)\n'
          + ' }',
      }],
    }
    const mockId = 'D:/project-github/unplugin-vue-cssvars/play/src/App.vue'
    const res = createCSSModule(mockDescriptor as any, mockId, mockCssFiles)
    expect(res).toMatchObject([mockCSSFilesContent])
    expect(res).matchSnapshot()
  })

  test('createCSSModule: multiple style tag', () => {
    const mockCssFiles = new Map()
    const mockCSSFilesContent = {
      importer: new Set(),
      vBindCode: {
        foo: new Set(['v-bind(foo)']),
      },
    }
    mockCssFiles.set('D:/project-github/unplugin-vue-cssvars/play/src/assets/test.css', mockCSSFilesContent)
    const mockCSSFilesContent2 = {
      importer: new Set(),
      vBindCode: {
        bar: new Set(['v-bind(bar)']),
      },
    }
    mockCssFiles.set('D:/project-github/unplugin-vue-cssvars/play/src/assets/test2.css', mockCSSFilesContent2)
    const mockDescriptor = {
      styles: [{
        content: '@import "./assets/test";\n'
          + ' div {\n'
          + '   color: v-bind(color)\n'
          + ' }',
      },
      {
        content: '@import "./assets/test2";\n'
          + ' div {\n'
          + '   color: v-bind(color2)\n'
          + ' }',
      }],
    }
    const mockId = 'D:/project-github/unplugin-vue-cssvars/play/src/App.vue'
    const res = createCSSModule(mockDescriptor as any, mockId, mockCssFiles)
    expect(res).toMatchObject([mockCSSFilesContent, mockCSSFilesContent2])
    expect(res).matchSnapshot()
  })

  test('createCSSModule: no style tag', () => {
    const mockCssFiles = new Map()
    mockCssFiles.set('foo', {
      importer: new Set(),
      vBindCode: {
        foo: new Set(['v-bind(foo)']),
      },
    })
    const mockDescriptor = {
      styles: [],
    }
    const res = createCSSModule(mockDescriptor as any, 'foo', mockCssFiles)
    expect(res.length).toBe(0)
  })
})
