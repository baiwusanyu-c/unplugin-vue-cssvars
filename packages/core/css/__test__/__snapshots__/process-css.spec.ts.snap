// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`process css > createCSSModule: basic 1`] = `
[
  {
    "importer": Set {},
    "vBindCode": {
      "foo": Set {
        "v-bind(foo)",
      },
    },
  },
]
`;

exports[`process css > createCSSModule: multiple style tag 1`] = `
[
  {
    "importer": Set {},
    "vBindCode": {
      "foo": Set {
        "v-bind(foo)",
      },
    },
  },
  {
    "importer": Set {},
    "vBindCode": {
      "bar": Set {
        "v-bind(bar)",
      },
    },
  },
]
`;

exports[`process css > createCSSModule: no file with lang 1`] = `
[
  {
    "importer": Set {},
    "vBindCode": {
      "foo": Set {
        "v-bind(foo)",
      },
    },
  },
]
`;
